File: 0240_Search_a_2D_Matrix_II/solution.java
package cn.huang.leetcode;

/*
240. Search a 2D Matrix II

Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

    Integers in each row are sorted in ascending from left to right.
    Integers in each column are sorted in ascending from top to bottom.

Example:

Consider the following matrix:

[
  [1,   4,  7, 11, 15],
  [2,   5,  8, 12, 19],
  [3,   6,  9, 16, 22],
  [10, 13, 14, 17, 24],
  [18, 21, 23, 26, 30]
]

Given target = 5, return true.

Given target = 20, return false.
Accepted
147,314
Submissions
370,564
 */
public class LC_0240_Searcha2DMatrix_II {
    /*
      bin search 0(m*LogN)
     */
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length < 1)
            return false;

        int rows = matrix.length;
        int cols = matrix[0].length;

        int row = 0, col = cols - 1;

        while (row < rows && col >= 0) {
            if (matrix[row][col] == target)
                return true;
            else if (matrix[row][col] < target)
                row++;
            else
                col--;
        }
        return false;
    }
}

File: 0240_Search_a_2D_Matrix_II/solution.cpp
/*
240. Search a 2D Matrix II

Level: Medium

https://leetcode.com/problems/search-a-2d-matrix-ii
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

class Solution
{
public:
  bool searchMatrix(vector<vector<int>> &matrix, int target)
  {
    if (matrix.empty() || matrix[0].empty())
      return false;
    const int m = matrix.size();
    const int n = matrix[0].size();
    int r = 0;
    int c = matrix[0].size() - 1;
    while (r < m && c >= 0)
    {
      if (matrix[r][c] == target)
        return true;
      else if (matrix[r][c] > target)
        --c;
      else
        ++r;
    }
    return false;
  }
};